tka/builder.go

```diff
@@ -0,0 +1,165 @@
+	meta  map[string]string
+	keyID KeyID
+}
+
+// UpdateBuilder implements a builder for changes to the tailnet
+// key authority.
+//
+// Finalize must be called to compute the update messages, which
+// must then be applied to all Authority objects using Inform().
+type UpdateBuilder struct {
+	a *Authority
+
+	keysToAdd     map[mapkeyKeyID]Key
+	votesToChange map[mapkeyKeyID]voteInfo
+	metaToChange  map[mapkeyKeyID]metaInfo
+	keysToRemove  map[mapkeyKeyID]KeyID
+}
+
+func mapKeyID(id KeyID) mapkeyKeyID {
+	var out mapkeyKeyID
+	copy(out[:], id)
+	return out
+}
+
+// AddKey adds a new key to the authority.
+func (b *UpdateBuilder) AddKey(key Key) {
+	b.keysToAdd[mapKeyID(key.ID())] = key
+}
+
+// RemoveKey removes a key from the authority.
+func (b *UpdateBuilder) RemoveKey(key KeyID) {
+	b.keysToRemove[mapKeyID(key)] = key
+}
+
+// SetKeyVote updates the number of votes of an existing key.
+func (b *UpdateBuilder) SetKeyVote(key KeyID, votes uint) {
+	b.votesToChange[mapKeyID(key)] = voteInfo{votes, key}
+}
+
+// SetKeyMeta updates key-value metadata stored against an existing key.
+//
+// TODO(tom): Provide an API to update specific values rather than the whole
+// map.
+func (b *UpdateBuilder) SetKeyMeta(key KeyID, meta map[string]string) {
+	b.metaToChange[mapKeyID(key)] = metaInfo{meta, key}
+}
+
+// Finalize computes the set of update message to actuate the update.
+// The provider signer function, if non-nil, is called with each update
+// to compute and apply signatures.
+func (b *UpdateBuilder) Finalize(signer func(*AUM) error) ([]AUM, error) {
+	var (
+		out    []AUM
+		state  = b.a.state
+		parent = b.a.Head()
+	)
+
+	// Helper function to commit an AUM to the output set.
+	makeUpdate := func(update AUM) error {
+		prevHash := make([]byte, len(parent))
+		copy(prevHash, parent[:])
+		update.PrevAUMHash = prevHash
+
+		if signer != nil {
+			if err := signer(&update); err != nil {
+				return fmt.Errorf("signing failed: %v", err)
+			}
+		}
+		if err := update.StaticValidate(); err != nil {
+			return fmt.Errorf("generated update was invalid: %v", err)
+		}
+
+		parent = update.Hash()
+		out = append(out, update)
+		return nil
+	}
+
+	// Generated AddKey message for new keys.
+	for _, ak := range b.keysToAdd {
+		if _, err := state.GetKey(ak.ID()); err == nil {
+			return nil, fmt.Errorf("cannot add key %v: already exists", ak)
+		}
+		err := makeUpdate(AUM{MessageKind: AUMAddKey, Key: &ak})
+		if err != nil {
+			return nil, fmt.Errorf("adding key %v: %v", ak, err)
+		}
+	}
+
+	// Generate UpdateKey messages for existing keys.
+	for k, info := range b.votesToChange {
+		if _, err := state.GetKey(info.keyID); err != nil {
+			return nil, fmt.Errorf("failed reading key %x: %v", info.keyID, err)
+		}
+		update := AUM{MessageKind: AUMUpdateKey, Votes: &info.votes, KeyID: info.keyID}
+		// If theres also a meta update for this key, roll it into one update.
+		if info, hasMetaUpdate := b.metaToChange[k]; hasMetaUpdate {
```

Comment:
I think this is the only reason for the builder fields to be maps, I wonder if it would be fine to just skip this optimization and use slices instead. Or even preserve it with slices -- yes that makes this a O(_n_<sup>2</sup>) operation but I expect _n_ to be small.

Or if you are doing these optimizations, I feel like there are others that are missing:
* Cancel out adds and deletes
* Cancel out vote/meta changes if there's a delete
* This covers [meta, vote]; also check for [vote, meta].

Also the vote/meta split could be resolved by making `SetKeyMeta` take an additional parameter for vote changes, which could be 0 or something like `const noVoteChange = 0`.

If you add the same operation twice (delete a particular key, change the vote for a particular key, etc), it'll silently overwrite the old instance, is that intentional?